controllers/operatorpolicy_controller.go

```diff
@@ -108,167 +104,345 @@ func (r *OperatorPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Reque
+			}
 
-			// Set to non compliant because operatorgroup does not exist on cluster, but
-			// should still go to check subscription
-			r.setCompliance(ctx, policy, policyv1.NonCompliant)
+			return nil
+		}
+
+		// check whether the specs match
+		desiredUnstruct, err := runtime.DefaultUnstructuredConverter.ToUnstructured(desiredOpGroup)
+		if err != nil {
+			return fmt.Errorf("error converting desired OperatorGroup to an Unstructured: %w", err)
+		}
+
+		updateNeeded, skipUpdate, err := r.mergeObjects(
+			ctx, desiredUnstruct, &opGroup, string(policy.Spec.ComplianceType),
+		)
+		if err != nil {
+			return fmt.Errorf("error checking if the OperatorGroup needs an update: %w", err)
+		}
+
+		if !updateNeeded {
+			// Everything relevant matches!
+			err := r.updateStatus(ctx, policy, matchesCond("OperatorGroup"), matchedObj(&opGroup))
+			if err != nil {
+				return fmt.Errorf("error updating the status for an OperatorGroup that matches: %w", err)
+			}
+
+			return nil
 		}
-	}
 
-	// Create new subscription
-	if cachedSubscription == nil {
-		if strings.EqualFold(string(policy.Spec.RemediationAction), string(policyv1.Enforce)) {
-			subscriptionSpec := buildSubscription(policy)
+		// Specs don't match.
 
-			err := r.Create(ctx, subscriptionSpec)
+		if policy.Spec.OperatorGroup == nil {
+			// The policy doesn't specify what the OperatorGroup should look like, but what is already
+			// there is not the default one the policy would create.
+			// FUTURE: check if the one operator group is compatible with the desired subscription.
+			// For an initial implementation, assume if an OperatorGroup already exists, then it's a good one.
+			err := r.updateStatus(ctx, policy, opGroupPreexistingCond, matchedObj(&opGroup))
 			if err != nil {
-				OpLog.Error(err, "Could not handle missing musthave object")
-				r.setCompliance(ctx, policy, policyv1.NonCompliant)
+				return fmt.Errorf("error updating the status for a pre-existing OperatorGroup: %w", err)
+			}
+
+			return nil
+		}
 
-				return reconcile.Result{}, err
+		if policy.Spec.RemediationAction.IsEnforce() && skipUpdate {
+			err = r.updateStatus(ctx, policy, mismatchCondUnfixable("OperatorGroup"), mismatchedObj(&opGroup))
+			if err != nil {
+				return fmt.Errorf("error updating status for an unenforceable mismatched OperatorGroup: %w", err)
 			}
 
-			// Future work: Check availability/status of all resources
-			// managed by the OperatorPolicy before setting compliance state
-			r.setCompliance(ctx, policy, policyv1.Compliant)
+			return nil
+		}
 
-			return reconcile.Result{}, nil
+		// The names match, but the specs don't: report NonCompliance
+		err = r.updateStatus(ctx, policy, mismatchCond("OperatorGroup"), mismatchedObj(&opGroup))
+		if err != nil {
+			return fmt.Errorf("error updating the status for an OperatorGroup that does not match: %w", err)
+		}
+
+		if policy.Spec.RemediationAction.IsEnforce() {
+			desiredOpGroup.ResourceVersion = opGroup.GetResourceVersion()
```

Comment:
Why newly updated opgroup should have the same resourceversion as old one?

Reply:
It won't really, but when sending an update request, the API requires what version you were looking at. That way, you don't submit an update that might conflict with something that changed in the meantime.

Reply:
oh I see Thanks!!